#### User management

junaid-work1 opened the pull request:
- Create api that user can register or create account.
- Create api that user can update the account.
- Create api that user can delete the account.
- Create api that user can get list of users.
- Create api that user can single user by id.
- Create api that user can sign in to account.
- Create separate Routes for each end point.
- Create generic function that return success and error response.

Commit:
#### User management

- Create api that user can register.
- Create api that user can update the account.
- Create api that user can delete the account.
- Create api that user can get list of users.
- Create api that user can single user by id.
- Create api that user can sign in to account.
- Create seperate Routes for each end point.
- Create generic function that return success and error response.

Commit:
- Separate the api response methods.
- Create separate method for generating jwt token.

Saim00786 commented on index.js:
```diff
@@ -3,6 +3,7 @@ const dotenv = require("dotenv");
 const cors = require("cors");
 const app = express();
 
+const userRoutes = require("./src/routes/UserRoutes");
```
Use import instead of require

Saim00786 commented on src/controllers/UserController.js:
```diff
@@ -0,0 +1,114 @@
+const { generateToken } = require("../helper/generateToken");
+
+
+exports.getAllUser = async (req, res) => {
```
```suggestion
exports.getAllUsers = async (req, res) => {
```

Saim00786 commented on src/controllers/UserController.js:
```diff
@@ -0,0 +1,114 @@
+      let newItem = item.toObject();
+      delete newItem.password;
+      return newItem;
+    });
```
Use database query to only select the required fields

Saim00786 commented on src/controllers/UserController.js:
```diff
@@ -0,0 +1,114 @@
+    const hashedPassword = await bcrypt.hash(password, 10);
+    
+    let newUser = new User({ ...req.body, password: hashedPassword });
+    let result = await newUser.save();
```
Use const instead of let

Saim00786 replied:
don't use req.body destructure

Saim00786 commented on src/controllers/UserController.js:
```diff
@@ -0,0 +1,114 @@
+  }
+};
+
+exports.getsingleUser = async (req, res) => {
```
```suggestion
exports.getSingleUser = async (req, res) => {
```

Saim00786 commented on src/controllers/UserController.js:
```diff
@@ -0,0 +1,114 @@
+  try {
+    const user = await User.findById(userId);
+    let result = user.toObject();
+    delete result.password;
```
What if user is null

Saim00786 commented on src/utils/helper.js:
```diff
@@ -0,0 +1,18 @@
+const sendSuccessResponse = (res, data, statusCode, statusText) => {
```
extra file

Saim00786 reviewed (commented):
Looks Good

Commit:
- Use import instead of require.
- Create separate api for updating password.
- Improve the logics of user controllers.

Commit:
Update user apis

- Add joi for data validation.
- Add feature of email verification.
- Change the schemas.
- Install Joi, Nodemailer, uuidv4 packages.
- Add milddeware for token verification.

Saim00786 commented on index.js:
```diff
@@ -1,14 +1,20 @@
-const dotenv = require("dotenv");
-const cors = require("cors");
-const app = express();
+import cors from 'cors'
```
👍🏻

Saim00786 commented on src/controllers/UserController.js:
```diff
@@ -0,0 +1,175 @@
+
+  const verificationCode = uuidv4();
+
+  const { error } = registerSchema.validate(req.body, { abortEarly: false });
```
```suggestion
  const { error } = registerUserSchema.validate(req.body, { abortEarly: false });
```

Saim00786 commented on src/controllers/UserController.js:
```diff
@@ -0,0 +1,175 @@
+    }
+    const user = await User.findOne({ email });
+    if (user) {
+      return sendErrorResponse(res, "User with this email already exists", 409, "fail");
```
```suggestion
      return sendErrorResponse(res, "User with this email already exists", 422, "fail");
```

Saim00786 commented on src/controllers/UserController.js:
```diff
@@ -0,0 +1,175 @@
+
+  try {
+    const updatedUser = await User.findByIdAndUpdate(userId, newData, {
+      new: true,
```
why we need to do this ?

Saim00786 reviewed (commented):
Other wise looks good

Commit:
- Update the routes index file
- Improve the routes
- Remove extra work
- Add pre hooks for remove data which are associated with user
